Update Boost Benefits interstitial page pricing

karthikax opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes 84-gh-Automattic/boost-shield-api

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Replaced hardcoded pricing in Jetpack Boost Benefits interstitial page with dynamically fetched pricing from the API.

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
NA

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to '..'
*

pyronaur commented on projects/plugins/boost/app/assets/src/js/global.d.ts:
```diff
@@ -48,6 +48,13 @@ declare global {
 		nonces: {
 			[ key: string ]: string;
 		};
+		pricing: {
```
This type doesn't match what `add_js_constants` produces.

karthikax replied:
I didn't get that. Do you mean the possibility of `yearly` being empty?

pyronaur replied:
Yes. We're not ensuring that `pricing` has any other values than `yearly`:

https://github.com/Automattic/jetpack/pull/24503/files#diff-d7bb61f231c4b097b8493b260a1dbd6e7d6b82b4d4972428a515efd3fdf03713R16-R18

Because it uses `Wpcom_Products::get_product_pricing` under the hood, and that function can return an empty array for example. Also because it's a function that we inherit via a package, we should also be prepared for a situation where the return format may change.

karthikax replied:
Makes sense. I have updated the type and how the constant is passed.

pyronaur commented on projects/plugins/boost/app/assets/src/js/pages/benefits/BenefitsInterstitial.svelte:
```diff
@@ -73,10 +75,10 @@
 					icon={`${ window.Jetpack_Boost.site.assetPath }../static/images/forward.svg`}
-					priceBefore={19.95}
-					priceAfter={9.95}
+					priceBefore={pricing.full_price / 12}
```
Does the react component handle weird float values like 1.3333333 or `0.1 + 0.2`?

karthikax replied:
I tested this with a few different values like 1.3333333, 0.1 + 0.2, '123', null, etc and the component handles it appropriately. Upon that, I think we should be sending predefined values to the UI anyway.

pyronaur replied:
Yeah, but `19.99 / 12 = 1.6658333333` so I' just wanted to make sure we're able to handle that :)

pyronaur reviewed (changes requested):
Looks like we're on the right path, but there are a couple of things to think about.

pyronaur commented:
How are we going to handle situations where the pricing information is missing?

karthikax commented:
> How are we going to handle situations where the pricing information is missing?

It would throw a console error earlier and not render the component. I have updated how we pass the constant to the UI and added a check before we render the pricing component.

pyronaur commented:
I think this is still missing an error message of some sort. 

The user is interested in purchasing boost, but because no pricing data is received, we're just not showing them any data, not even a "Get in touch" or "Try again".

Speaking of which - are we logging the intent and failure?

karthikax commented:
That looks like a good idea. Since that involves a design decision I have created a new ticket for the same: 88-gh-Automattic/boost-shield-api. I have added 3 possible solutions. I think solutions 1 and 3 are easier to implement, solution 2 would gives a better user experience.

karthikax commented:
Following the conversation in 88-gh-Automattic/boost-shield-api and p1653651077157419-slack-C016BBAFHHS we will only log a Tracks event and redirect the user directly to the checkout page if the price information is missing.
I have updated this PR accordingly

thingalon commented on projects/plugins/boost/app/assets/src/js/utils/analytics.ts:
```diff
@@ -1,18 +1,38 @@
+
+export async function recordBoostEvent(
 	eventName: string,
 	eventType: string,
```
We're not using the `eventType` argument here. Doesn't have to be a part of this PR, but it would be nice to eliminate this unused argument across the project. Should be a 2 minute search & modify.

thingalon reviewed (commented):
Tested, and appears to work. The code in here seems ok to me. I didn't know about the `// <reference` syntax, but it seems like a good way to avoid jQuery types creeping further into our code.

Preliminary 👍 from me but I'll give Nauris another opportunity to look.

pyronaur reviewed (approved):
🚢
